### ISSUE TITLE
c-callable not correct in LLVM back-end

### ISSUE THREAD

**Issue by waywardmonkeys** [dylan-lang/opendylan#1018]
In the ``objective-c`` library, I use ``C-callable`` to implement methods in the Objective C object system with code written in Dylan.

When compiled with the LLVM back-end, this results in a segfault at runtime.

The LLVM IR for this looks like:

```llvm
define internal fastcc %struct.MV @c_callable_adder_c_wrapper_(i8* %targetF1, i8* %selectorF2, i32 %aF3, i8* %.next, i8* %.function) {
bb.entry:
  call void @llvm.dbg.value(metadata !{i8* %targetF1}, i64 0, metadata !286) #0, !dbg !287
  call void @llvm.dbg.value(metadata !{i8* %selectorF2}, i64 0, metadata !288) #0, !dbg !287
  call void @llvm.dbg.value(metadata !{i32 %aF3}, i64 0, metadata !289) #0, !dbg !287
  call void @llvm.dbg.value(metadata !{i8* %.next}, i64 0, metadata !290) #0, !dbg !287
  call void @llvm.dbg.value(metadata !{i8* %.function}, i64 0, metadata !291) #0, !dbg !287
  %0 = alloca %KLsimple_object_vectorGVKd.2, align 4, !dbg !287
  %1 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %0, i32 0, i32 0, !dbg !287
  store %KLmm_wrapperGVKi* @KLsimple_object_vectorGVKdW, %KLmm_wrapperGVKi** %1, align 1, !dbg !287
  %2 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %0, i32 0, i32 1, !dbg !287
  store i8* inttoptr (i32 9 to i8*), i8** %2, align 1, !dbg !287
  %3 = alloca %KLsimple_object_vectorGVKd.2, align 4, !dbg !287
  %4 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %3, i32 0, i32 0, !dbg !287
  store %KLmm_wrapperGVKi* @KLsimple_object_vectorGVKdW, %KLmm_wrapperGVKi** %4, align 1, !dbg !287
  %5 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %3, i32 0, i32 1, !dbg !287
  store i8* inttoptr (i32 9 to i8*), i8** %5, align 1, !dbg !287
  %6 = ptrtoint i8* %targetF1 to i32, !dbg !292
  %7 = call fastcc %KLmachine_wordGVKe* @primitive_wrap_machine_word(i32 %6), !dbg !292
  %8 = bitcast %KLmachine_wordGVKe* %7 to i8*, !dbg !292
  %9 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %0, i32 0, i32 2, i32 0, !dbg !292
  %10 = call {}* @llvm.invariant.start(i64 4, i8* nocapture bitcast (i8** @IKJaddress_ to i8*)) #1, !dbg !292
  %11 = load i8** @IKJaddress_, align 4, !dbg !292
  store i8* %11, i8** %9, align 4, !dbg !292
  %12 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %0, i32 0, i32 2, i32 1, !dbg !292
  store i8* %8, i8** %12, align 4, !dbg !292
  %13 = bitcast %KLsimple_object_vectorGVKd.2* %0 to i8*, !dbg !292
  %14 = bitcast i8* bitcast (%KLincremental_generic_functionGVKe* @KmakeVKd to i8*) to %KLgeneric_functionGVKd*, !dbg !292
  %15 = getelementptr inbounds %KLgeneric_functionGVKd* %14, i32 0, i32 6, !dbg !292
  %16 = load i8** %15, align 4, !dbg !292
  %17 = bitcast i8* %16 to %KLengine_nodeGVKg*, !dbg !292
  %18 = getelementptr inbounds %KLengine_nodeGVKg* %17, i32 0, i32 3, !dbg !292
  %19 = load i8** %18, align 4, !dbg !292
  %20 = bitcast i8* %19 to %struct.MV (i8*, i8*, i8*, i8*)*, !dbg !292
  %21 = call %struct.MV %20(i8* %16, i8* bitcast (%KLincremental_generic_functionGVKe* @KmakeVKd to i8*), i8* bitcast (%KLc_mapped_designator_classGVKi* @KLobjcSinstanceGVobjective_c to i8*), i8* %13), !dbg !292
  %22 = extractvalue %struct.MV %21, 0, !dbg !292
  %23 = call fastcc %struct.MV @Kimport_function_for_LobjcSinstanceGVobjective_cI(i8* %22, i8* undef, i8* undef), !dbg !293
  %24 = extractvalue %struct.MV %23, 0, !dbg !293
  %25 = ptrtoint i8* %selectorF2 to i32, !dbg !292
  %26 = call fastcc %KLmachine_wordGVKe* @primitive_wrap_machine_word(i32 %25), !dbg !292
  %27 = bitcast %KLmachine_wordGVKe* %26 to i8*, !dbg !292
  %28 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %3, i32 0, i32 2, i32 0, !dbg !292
  %29 = call {}* @llvm.invariant.start(i64 4, i8* nocapture bitcast (i8** @IKJaddress_ to i8*)) #1, !dbg !292
  %30 = load i8** @IKJaddress_, align 4, !dbg !292
  store i8* %30, i8** %28, align 4, !dbg !292
  %31 = getelementptr inbounds %KLsimple_object_vectorGVKd.2* %3, i32 0, i32 2, i32 1, !dbg !292
  store i8* %27, i8** %31, align 4, !dbg !292
  %32 = bitcast %KLsimple_object_vectorGVKd.2* %3 to i8*, !dbg !292
  %33 = bitcast i8* bitcast (%KLincremental_generic_functionGVKe* @KmakeVKd to i8*) to %KLgeneric_functionGVKd*, !dbg !292
  %34 = getelementptr inbounds %KLgeneric_functionGVKd* %33, i32 0, i32 6, !dbg !292
  %35 = load i8** %34, align 4, !dbg !292
  %36 = bitcast i8* %35 to %KLengine_nodeGVKg*, !dbg !292
  %37 = getelementptr inbounds %KLengine_nodeGVKg* %36, i32 0, i32 3, !dbg !292
  %38 = load i8** %37, align 4, !dbg !292
  %39 = bitcast i8* %38 to %struct.MV (i8*, i8*, i8*, i8*)*, !dbg !292
  %40 = call %struct.MV %39(i8* %35, i8* bitcast (%KLincremental_generic_functionGVKe* @KmakeVKd to i8*), i8* bitcast (%KLc_mapped_designator_classGVKi* @KLobjcSselectorGVobjective_c to i8*), i8* %32), !dbg !292
  %41 = extractvalue %struct.MV %40, 0, !dbg !292
  %42 = call fastcc i8* @primitive_wrap_abstract_integer(i32 %aF3), !dbg !294
  %43 = bitcast i8* bitcast (%KLunionGVKe* @KLffi_integerGYc_ffi_interfaceVc_ffi to i8*) to %KLtypeGVKd*, !dbg !294
  %44 = getelementptr inbounds %KLtypeGVKd* %43, i32 0, i32 1, !dbg !294
  %45 = load i8** %44, align 4, !dbg !294
  %46 = bitcast i8* %45 to %struct.MV (i8*, i8*, i8*, i8*)*, !dbg !294
  %47 = call fastcc %struct.MV %46(i8* %42, i8* bitcast (%KLunionGVKe* @KLffi_integerGYc_ffi_interfaceVc_ffi to i8*), i8* undef, i8* undef), !dbg !294
  %48 = extractvalue %struct.MV %47, 0, !dbg !294
  %49 = icmp ne i8* %48, bitcast (%KLbooleanGVKd* @KPfalseVKi to i8*), !dbg !294
  %50 = call i1 @llvm.expect.i1(i1 %49, i1 true) #0, !dbg !294
  br i1 %49, label %53, label %51, !dbg !294

; <label>:51                                      ; preds = %bb.entry
  %52 = tail call fastcc %struct.MV @Ktype_check_errorVKiI(i8* %42, i8* bitcast (%KLunionGVKe* @KLffi_integerGYc_ffi_interfaceVc_ffi to i8*), i8* undef, i8* undef), !dbg !294
  unreachable, !dbg !294

; <label>:53                                      ; preds = %bb.entry
  %54 = bitcast i8* bitcast (%KLsimple_methodGVKe* @KadderVobjc_test_suite to i8*) to %KLfunctionGVKd*, !dbg !292
  %55 = getelementptr inbounds %KLfunctionGVKd* %54, i32 0, i32 1, !dbg !292
  %56 = load i8** %55, align 4, !dbg !292
  %57 = bitcast i8* %56 to %struct.MV (i8*, i32, i8*, i8*, i8*)*, !dbg !292
  %58 = call %struct.MV %57(i8* bitcast (%KLsimple_methodGVKe* @KadderVobjc_test_suite to i8*), i32 3, i8* %24, i8* %41, i8* %42), !dbg !292
  %59 = extractvalue %struct.MV %58, 0, !dbg !292
  %60 = call fastcc i32 @primitive_unwrap_abstract_integer(i8* %59), !dbg !294
  ret i32 %60, !dbg !292
}
```

It gets used at a callsite:

```llvm
  %3 = ptrtoint i8* bitcast (%struct.MV (i8*, i8*, i32, i8*, i8*)* @c_callable_adder_c_wrapper_ to i8*) to i32, !dbg !768
```

**Comment by waywardmonkeys** [comment]
For what it is worth, my original code in Dylan looks like:

```dylan
define objc-method adder
    (target, selector, a :: <integer>)
 => (r :: <integer>)
  c-signature: (target :: <objc/instance>,
                selector :: <objc/selector>,
                a :: <C-int>) => (r :: <C-int>);
  assert-true(instance?(target, <test-class>));
  a + 1
end;
```

So that generated LLVM IR above is really pretty bulky looking.

**Comment by waywardmonkeys** [comment]
FWIW, there's no specialized handling of ``<&c-callable-function>`` in the LLVM back-end.

**Comment by waywardmonkeys** [comment]
Fixed by @housel in fc6f4e9ac4db65ab1cce395c360601f1d204446c.

(Confirmed by running my `objective-c` test suite.)

Thanks!
